File: Helper/content_search.py
#!/usr/bin/env python3
import os
import gzip
import sys

def name_match(filename: str, name_filter: str) -> bool:
    if not name_filter:
        return True
    return name_filter.lower() in filename.lower()

def content_in_text_file(path: str, needle: str) -> bool:
    try:
        with open(path, "r", encoding="utf-8", errors="ignore") as f:
            for line in f:
                if needle in line:
                    return True
    except Exception:
        return False
    return False

def content_in_gz_file(path: str, needle: str) -> bool:
    try:
        with gzip.open(path, "rt", encoding="utf-8", errors="ignore") as f:
            for line in f:
                if needle in line:
                    return True
    except Exception:
        return False
    return False

def main():
    base_dir = input("1) Directory: ").strip()
    if not os.path.isdir(base_dir):
        print(f"ERROR: Directory not found: {base_dir}")
        sys.exit(1)

    name_filter = input("2) File name filter (optional, press Enter to skip): ").strip()
    needle = input("3) Content to search (word/sentence): ").strip()
    if not needle:
        print("ERROR: Content keyword cannot be empty.")
        sys.exit(1)

    print("\nResult:\n-------------------")
    results = []

    for root, _, files in os.walk(base_dir):
        for fn in files:
            if not name_match(fn, name_filter):
                continue

            path = os.path.join(root, fn)

            if fn.endswith(".gz"):
                if content_in_gz_file(path, needle):
                    results.append(path)
            else:
                if content_in_text_file(path, needle):
                    results.append(path)

    for p in results:
        print(p)

    print(f"\nSummary\n{base_dir.rstrip('/')}/ : {len(results)} files")

if __name__ == "__main__":
    main()

File: Helper/content_search.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Find;
use IO::Uncompress::Gunzip qw(gunzip $GunzipError);

print "1) Directory: ";
chomp(my $base_dir = <STDIN>);
die "ERROR: Directory not found: $base_dir\n" if !-d $base_dir;

print "2) File name filter (optional, press Enter to skip): ";
chomp(my $name_filter = <STDIN>);
my $nf = lc($name_filter);

print "3) Content to search (word/sentence): ";
chomp(my $needle = <STDIN>);
die "ERROR: Content keyword cannot be empty.\n" if !$needle;

my @results;

sub name_match {
    my ($filename) = @_;
    return 1 if $name_filter eq "";
    return index(lc($filename), $nf) >= 0;
}

sub file_contains {
    my ($path) = @_;
    if ($path =~ /\.gz$/) {
        my $gz = IO::Uncompress::Gunzip->new($path);
        return 0 if !$gz;
        while (my $line = <$gz>) {
            if (index($line, $needle) >= 0) { $gz->close(); return 1; }
        }
        $gz->close();
        return 0;
    } else {
        open my $fh, "<", $path or return 0;
        while (my $line = <$fh>) {
            if (index($line, $needle) >= 0) { close $fh; return 1; }
        }
        close $fh;
        return 0;
    }
}

find(sub {
    return if !-f $_;
    return if !name_match($_);
    my $path = $File::Find::name;
    push @results, $path if file_contains($path);
}, $base_dir);

print "\nResult:\n-------------------\n";
print "$_\n" for @results;
print "\nSummary\n" . ($base_dir =~ s{/$}{}r) . "/ : " . scalar(@results) . " files\n";